### PR TITLE
Support API Resources

### PR DESCRIPTION
**Pull request by winter-ice**
resolve https://github.com/inertiajs/inertia-laravel/issues/27 https://github.com/inertiajs/inertia-laravel/issues/28
Use like:
```
$user = UserResource::make(User::find(1));
return Inertia::render('Users/Edit', $user)
```
```
$users = UserResource::collection(User::paginate());
return Inertia::render('Users/Index', $users)
```
or
```
$users = UserCollection::make(User::paginate());
return Inertia::render('Users/Index', $users);
```

### CODE REVIEW

**Comment by reinink** [comment]
Hmm, I thought this worked already? I've used API resources. They automatically get converted to JSON, so this shouldn't be needed.

Plus, we also added `Arrayable` support recently (although it's not tagged yet): #39

I'll tag that now. Can you try this again, on the latest version? I suspect it will work without any changes needed to the library.

**Comment by winter-ice** [comment]
1. Passed to render function is an instance of `JsonResource`,Not instance of `Arrayable`.So I think it's not work.
2. Before response,The data not be fully handled.(such as `meta`,`link`).

forgive my poor English 😄

**Comment by winter-ice** [comment]
And I have test many usage.Here is some examples:
```
$user = User::find(1);
$props = UserResource::make($user);

return Inertia::render('Users/Edit', $props)
```
```
$user = User::find(1);
$props = UserResource::make($user)->additional(['foo'=>'bar']);

return Inertia::render('Users/Edit', $props)
```
And the `with` method in your resource class is works well.

Notice that if you want to append other data to a collection.
there are two ways:
1. use `additional`
2. defined `with` method in your resource collection.

defined in a resource and use `UserResource::collection` is not works

**Comment by sebastiandedeyne** [comment]
> Before response,The data not be fully handled.(such as meta,link).

This is true, and a super annoying thing with Laravel resources. When calling `toArray` you lose meta and pagination data.

It's debatable whether Inertia should cast this for the user though, since some people might expect to only get the data when passing down an Eloquent resource collection.

I personally would be very happy with this change, but I don't know what others are doing with resources; it might be to opinionated.

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,5 @@
 /vendor
+.idea
```
Comment (m1guelpf) [inline .gitignore]: This change seems unrelated to the PR

Comment (winter-ice) [reply .gitignore]: Yes,It's unrelated to the PR,I just get used to add this.

Comment (sebastiandedeyne) [reply .gitignore]: Editor settings should be globally ignored :) https://help.github.com/en/articles/ignoring-files#create-a-global-gitignore

Comment (winter-ice) [reply .gitignore]: Thanks,I don't know that.

Comment (m1guelpf) [inline .gitignore]:
```suggestion
```

**Comment by thoresuenert** [comment]
This PR will save time and a Collection class if you go the Laravel resource route.

In our team the resource is a good api for transforming data, I like it a lot.

File: src/ResponseFactory.php [hunk 2]
```diff
@@ -48,6 +49,10 @@ public function getVersion()
 
     public function render($component, $props = [])
     {
+        if ($props instanceof JsonResource) {
```
Comment (grantholle) [inline src/ResponseFactory.php]: So in my case, I would be passing an array of props, with some being an instance of `JsonResource`.

It would need to iterate each of the props to resolve them as well. It may be worth rethinking this a little...

**Comment by lexdewilligen** [comment]
@reinink Any updates on this? I need the resource meta, so I'd love to see one of the solutions merged..

**Comment by reinink** [comment]
So, I've added support for "Responsable" props: 2b966ac27964534ee3bbe585b0ea0fff8babb53a

As you can see, my approach here is almost identical to this PR. However, instead of checking for a `JsonResource` object, I'm checking for the more generic `Responsable` object. This makes it useful in other situations as well.

I've tested this with both single resource responses and collection resource responses, and it works great.

Thanks, either way, for submitting this PR @winter-ice. 👍
